pyracf/common/logger.py

```diff
@@ -93,33 +95,63 @@ def log_debug(self, header_message: str, message: str) -> None:
+    def __redact_request_dictionary(
         self,
-        string: Union[str, bytes],
-        redact_strings: List[str],
-    ) -> str:
-        """Redact a list of strings or sequences of bytes in a string or bytes object"""
-        if not redact_strings:
-            return string
-        for to_redact in redact_strings:
-            if isinstance(string, bytes):
-                string = self.redact_string(
-                    string, redact_string=bytes(to_redact, "utf-8")
-                )
-            else:
-                string = self.redact_string(string, redact_string=to_redact)
-        return string
+        dictionary: dict,
+        secret_traits: dict,
+    ) -> dict:
+        """Redact a list of specified secret traits in a dictionary"""
+        for seg_trait in secret_traits:
+            segment = seg_trait.split(":")[0]
+            try:
+                if not dictionary.get(segment, {}).get(seg_trait, {}).get("value", ""):
+                    continue
+                dictionary[segment][seg_trait]["value"] = "********"
+            except AttributeError:
+                continue
+        return dictionary
+
+    def __redact_string(
+        self,
+        input_string: Union[str, bytes],
+        start_ind: int,
+        end_pattern: str,
+    ):
+        pre_keyword = input_string[:start_ind]
+        post_keyword = end_pattern.join(input_string[start_ind:].split(end_pattern)[1:])
+        return pre_keyword + "********" + end_pattern + post_keyword
 
-    def redact_string(
-        self, string: Union[str, bytes], redact_string: Union[str, bytes]
+    def redact_request_xml(
+        self,
+        xml_string: Union[str, bytes],
+        secret_traits: dict,
+    ) -> Union[str, bytes]:
+        """Redact a list of specifid secret traits in a request xml string or bytes object"""
+        is_bytes = False
+        if isinstance(xml_string, bytes):
+            is_bytes = True
+            xml_string = xml_string.decode("utf-8")
+        for xml_key in secret_traits.values():
+            match = re.search(r"\<{}+[^>]*\>".format(xml_key), xml_string)
+            if not match:
+                continue
+            xml_string = self.__redact_string(xml_string, match.end(), "</")
+        if is_bytes:
+            xml_string = xml_string.encode("utf-8")
+        return xml_string
+
+    def redact_result_xml(
+        self,
+        xml_string: str,
+        secret_traits: dict,
     ) -> str:
-        """Redact a string or sequence of byte in a bytes object."""
-        if not redact_string:
-            return string
-        redacted_string = "********"
-        if isinstance(redact_string, bytes):
-            redacted_string = b"********"
-        return string.replace(redact_string, redacted_string)
+        for xml_key in secret_traits.values():
+            racf_key = xml_key.split(":")[1] if ":" in xml_key else xml_key
+            match = re.search(r"{} +\(".format(racf_key.upper()), xml_string)
```

Comment:
Does the uppercase version of racf_key always correspond to what shows up in the returned TSO commands? It seems like it does, but I am just double checking. 

Also, I think using this syntax for your regexes is a little cleaner:

```python
            match = re.search(f"\<{xml_key}+[^>]*\>", xml_string)
```

```python
            match = re.search(f"{racf_key.upper()} +\(", xml_string)
```

I tested this out and I believe this syntax is functionally the same. I am not a fan of using `.format` because it add more clutter in my opinion.

Reply:
Changed again to `rf` strings to avoid flake8 errors. Still passes all unit tests.